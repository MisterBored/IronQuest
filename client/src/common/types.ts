export enum ActionType {
  LAMP = 'LAMP',
  QUEST = 'QUEST',
  TRAIN = 'TRAIN',
}

export enum Skill {
  AGILITY = 'AGILITY',
  ATTACK = 'ATTACK',
  CONSTITUTION = 'CONSTITUTION',
  CONSTRUCTION = 'CONSTRUCTION',
  COOKING = 'COOKING',
  CRAFTING = 'CRAFTING',
  DEFENCE = 'DEFENCE',
  DIVINATION = 'DIVINATION',
  DUNGEONEERING = 'DUNGEONEERING',
  FARMING = 'FARMING',
  FIREMAKING = 'FIREMAKING',
  FISHING = 'FISHING',
  FLETCHING = 'FLETCHING',
  HERBLORE = 'HERBLORE',
  HUNTER = 'HUNTER',
  INVENTION = 'INVENTION',
  MAGIC = 'MAGIC',
  MINING = 'MINING',
  PRAYER = 'PRAYER',
  RANGED = 'RANGED',
  RUNECRAFTING = 'RUNECRAFTING',
  SLAYER = 'SLAYER',
  SMITHING = 'SMITHING',
  STRENGTH = 'STRENGTH',
  SUMMONING = 'SUMMONING',
  THIEVING = 'THIEVING',
  WOODCUTTING = 'WOODCUTTING',
}

export interface Player {
  name: string;
  levels: Map<Skill, number>;
  questPoints: number;
  totalLevel: number;
  combatLevel: number;
}

export interface Action {
  type: ActionType;
  player: Player;
  future: boolean;
  message: string;
}

export interface Quest {
  displayName: string;
}

export interface LampAction extends Action {
  quest: Quest;
}

export interface QuestAction extends Action {
  quest: Quest;
}

export type TrainAction = Action;

export interface PathStats {
  percentComplete: number;
}

export interface Path {
  actions: Action[];
  stats: PathStats;
}

export enum QuestAccessFilter {
  ALL = 'ALL',
  FREE = 'FREE',
  MEMBERS = 'MEMBERS',
}

export enum QuestPriority {
  MAXIMUM = 'MAXIMUM',
  HIGH = 'HIGH',
  NORMAL = 'NORMAL',
  LOW = 'LOW',
  MINIMUM = 'MINIMUM',
}

export enum QuestTypeFilter {
  ALL = 'ALL',
  QUESTS = 'QUESTS',
  SAGAS = 'SAGAS',
  MINIQUESTS = 'MINIQUESTS',
}

export enum PathFinderAlgorithm {
  DEFAULT = 'DEFAULT',
  SMART_PRIORITIES = 'SMART_PRIORITIES',
}

export interface QuestPriorities {
  [key: number]: QuestPriority;
}

export interface QuestsParameters {
  name?: string;
  accessFilter?: QuestAccessFilter;
  typeFilter?: QuestTypeFilter;
}

export interface PathFinderParameters extends QuestsParameters {
  ironman?: boolean;
  recommended?: boolean;
  lampSkills?: Skill[];
  questPriorities?: QuestPriorities;
  algorithm?: PathFinderAlgorithm;
}

export interface Quest {
  id: number;
  displayName: string;
  priority: QuestPriority;
}

export interface AjaxError {
  response: Response;
  parameters: QuestsParameters | PathFinderParameters;
}
