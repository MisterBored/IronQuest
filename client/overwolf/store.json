{
  "category_description": "for games",
  "tags": ["runescape", "rs", "rs3", "iron", "quest"],
  "title": "IronQuest"
}
