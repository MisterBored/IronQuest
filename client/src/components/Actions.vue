<template>
  <v-container v-if="loading" fluid fill-height>
    <v-row justify="center" align="center">
      <v-progress-circular indeterminate color="primary" />
    </v-row>
  </v-container>
  <v-container v-else fluid>
    <v-row v-if="path">
      <v-col>
        <v-chip color="green">
          <v-avatar left>
            <v-icon>{{ mdiCheckboxMarkedCircle }}</v-icon>
          </v-avatar>
          Completed: {{ path.stats.percentComplete }}%
        </v-chip>
      </v-col>
      <v-col>
        <slot />
      </v-col>
    </v-row>
    <v-row v-if="path && !path.actions">
      <p>None</p>
    </v-row>
    <v-row v-if="path && path.actions">
      <v-col cols="12" md="6" lg="6" xl="6">
        <v-list class="actions-list" dense>
          <v-list-item-group v-model="selectedAction" mandatory>
            <v-list-item
              v-for="(action, i) in path.actions"
              :key="i"
              :value="action"
            >
              <v-list-item-icon class="mr-4">
                <v-icon>{{ getActionIcon(action) }}</v-icon>
              </v-list-item-icon>

              <v-list-item-content>
                <v-list-item-title v-text="action.message" />
              </v-list-item-content>
            </v-list-item>
          </v-list-item-group>
        </v-list>
      </v-col>
      <v-col
        v-if="selectedAction"
        cols="12"
        md="6"
        lg="6"
        xl="6"
        class="actions-content"
      >
        <v-simple-table dense>
          <tbody>
            <tr v-for="(row, index) in SKILLS_TABLE" :key="index">
              <template v-for="cell in row">
                <td :key="`${cell}-name`">
                  {{ capitalize(cell) }}
                </td>
                <td :key="`${cell}-level`">
                  {{ get(selectedAction.player.levels, cell) }}
                </td>
              </template>
            </tr>
            <tr>
              <td>Total level</td>
              <td>
                {{ selectedAction.player.totalLevel }}
              </td>
              <td>Combat level</td>
              <td>
                {{ selectedAction.player.combatLevel }}
              </td>
              <td>Quest points</td>
              <td>
                {{ selectedAction.player.questPoints }}
              </td>
            </tr>
            <tr v-if="showViewQuest(selectedAction)">
              <td colspan="6" class="view-quest">
                <a
                  :href="getActionQuestUrl(selectedAction)"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  View quest on wiki
                </a>
              </td>
            </tr>
          </tbody>
        </v-simple-table>
      </v-col>
    </v-row>
    <v-row v-if="error">
      <v-col>
        <slot />
      </v-col>
    </v-row>
    <v-row justify="center" align="center">
      <v-alert
        :value="error"
        type="error"
        transition="fade-transition"
        dismissible
        prominent
      >
        <v-row align="center">
          <v-col class="grow"> Something went wrong! </v-col>
          <v-col class="shrink">
            <v-btn
              :href="NEW_ISSUE_URL"
              target="_blank"
              rel="noopener noreferrer"
            >
              Submit issue
            </v-btn>
          </v-col>
        </v-row>
        <v-row v-if="errorResponse" align="center">
          <v-col>
            <v-expansion-panels>
              <v-expansion-panel>
                <v-expansion-panel-header>
                  View details
                </v-expansion-panel-header>
                <v-expansion-panel-content>
                  Parameters:
                  <pre>{{ errorResponse.parameters }}</pre>
                  Response:
                  <pre>{{ errorResponse.response }}</pre>
                </v-expansion-panel-content>
              </v-expansion-panel>
            </v-expansion-panels>
          </v-col>
        </v-row>
      </v-alert>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue';
import {
  Action,
  ActionType,
  LampAction,
  QuestAction,
  Skill,
} from '../common/types';
import {capitalize, get} from 'lodash';
import {mapFields} from 'vuex-map-fields';
import {ComputedMapper, RootState} from '../store';
import {
  mdiArrowUpBoldCircle,
  mdiChartBar,
  mdiCheckboxMarkedCircle,
  mdiCompassRose,
} from '@mdi/js';

const NEW_ISSUE_URL = 'https://github.com/darrenswhite/ironquest/issues/new';
const RUNESCAPE_WIKI_URL = 'https://runescape.wiki/';
const SKILLS_TABLE = [
  [Skill.ATTACK, Skill.CONSTITUTION, Skill.MINING],
  [Skill.STRENGTH, Skill.AGILITY, Skill.SMITHING],
  [Skill.DEFENCE, Skill.HERBLORE, Skill.FISHING],
  [Skill.RANGED, Skill.THIEVING, Skill.COOKING],
  [Skill.PRAYER, Skill.CRAFTING, Skill.FIREMAKING],
  [Skill.MAGIC, Skill.FLETCHING, Skill.WOODCUTTING],
  [Skill.RUNECRAFTING, Skill.SLAYER, Skill.FARMING],
  [Skill.CONSTRUCTION, Skill.HUNTER, Skill.SUMMONING],
  [Skill.DUNGEONEERING, Skill.DIVINATION, Skill.INVENTION],
];

export default Vue.extend({
  name: 'Actions',
  computed: {
    NEW_ISSUE_URL: () => NEW_ISSUE_URL,
    SKILLS_TABLE: () => SKILLS_TABLE,
    mdiCheckboxMarkedCircle: () => mdiCheckboxMarkedCircle,
    ...(mapFields({
      error: 'actions.error',
      errorResponse: 'actions.errorResponse',
      loading: 'actions.loading',
      path: 'actions.path',
      selectedAction: 'actions.selectedAction',
    }) as ComputedMapper<RootState['actions']>),
  },
  methods: {
    capitalize,
    get,
    getActionIcon(action: Action): string {
      let icon = null;

      switch (action.type) {
        case ActionType.LAMP:
          icon = mdiChartBar;
          break;
        case ActionType.QUEST:
          icon = mdiCompassRose;
          break;
        case ActionType.TRAIN:
          icon = mdiArrowUpBoldCircle;
          break;
      }

      return icon;
    },
    getActionQuestUrl(action: LampAction | QuestAction): string {
      return RUNESCAPE_WIKI_URL + action.quest.displayName.replace(/ /g, '_');
    },
    showViewQuest(action: Action): boolean {
      return (
        action.type === ActionType.QUEST || action.type === ActionType.LAMP
      );
    },
  },
});
</script>

<style lang="scss" scoped>
pre {
  white-space: pre-wrap;
}

.actions-list {
  max-height: 16rem;
  overflow-y: auto;
}

.actions-content {
  overflow-x: auto;
}

.view-quest {
  text-align: center;
}
</style>
