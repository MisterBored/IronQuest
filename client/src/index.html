<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      content="width=device-width, initial-scale=1, shrink-to-fit=no"
      name="viewport"
    />
    <meta content="IronQuest" name="description" />
    <meta content="Darren S. White" name="author" />
    <meta name="theme-color" content="#121212" />

    <link
      href="https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900"
      rel="stylesheet"
    />
    <link rel="manifest" href="/manifest.json" />
    <link rel="apple-touch-icon" href="/assets/icon-192x192.png" />

    <title>IronQuest</title>
  </head>

  <body>
    <noscript>
      To run this application, JavaScript is required to be enabled.
    </noscript>
    <div id="app"></div>
  </body>
</html>
