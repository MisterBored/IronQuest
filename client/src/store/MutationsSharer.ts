import createMutationsSharer, {
  BroadcastChannelStrategy,
  LocalStorageStratery,
  Strategy,
} from 'vuex-shared-mutations';

import {RootState} from './RootState';

function getStrategy(): Strategy {
  let strategy;

  if (BroadcastChannelStrategy.available()) {
    strategy = new BroadcastChannelStrategy({
      key: 'ironquest',
    });
  }

  if (LocalStorageStratery.available()) {
    strategy = new LocalStorageStratery({
      key: 'ironquest',
    });
  }

  if (!strategy) {
    throw new Error('No strategies available');
  }

  return strategy;
}

const mutationsSharer = createMutationsSharer<RootState>({
  predicate: () => true,
  strategy: getStrategy(),
});

export {mutationsSharer};
