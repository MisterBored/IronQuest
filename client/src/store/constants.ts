export const GET_FIELD = 'getField';
export const GET_PARAMETERS = 'getParameters';
export const FIND_PATH = 'findPath';
export const SET_ERROR = 'setError';
export const SET_PARAMETERS = 'setParameters';
export const SET_PATH = 'setPath';
export const SHOW_LOADER = 'showLoader';
export const UPDATE_FIELD = 'updateField';
