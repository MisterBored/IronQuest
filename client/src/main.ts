import Vue, {VueConstructor} from 'vue';
import Vuetify from 'vuetify/lib';
import colors from 'vuetify/lib/util/colors';

import {store} from './store';

import './styles/base.scss';

Vue.config.productionTip = false;

Vue.use(Vuetify);

const vuetify = new Vuetify({
  icons: {
    iconfont: 'mdiSvg',
  },
  theme: {
    dark: true,
    themes: {
      dark: {
        primary: colors.amber.darken3,
      },
    },
  },
});

export function createVue(app: VueConstructor<Vue>): Vue {
  return new Vue({
    store,
    vuetify,
    render: h => h(app),
  }).$mount('#app');
}
