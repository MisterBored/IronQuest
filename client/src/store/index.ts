import axios from 'axios';
import {extend, filter, head, map} from 'lodash';
import qs from 'qs';
import Vue from 'vue';
import Vuex, {
  ActionContext,
  ActionTree,
  GetterTree,
  MutationTree,
  Store,
} from 'vuex';
import {getField, updateField} from 'vuex-map-fields';

import {
  AjaxError,
  Path,
  PathFinderAlgorithm,
  PathFinderParameters,
  Quest,
  QuestAccessFilter,
  QuestPriority,
  QuestTypeFilter,
} from '../common/types';

import constants from './constants';
import {plugins} from './plugins';
import {RootState} from './RootState';

const QUESTS_URL = __API__ + '/quests';
const PATH_FINDER_URL = QUESTS_URL + '/path';

Vue.use(Vuex);

const state: RootState = {
  actions: {
    error: false,
    errorResponse: undefined,
    loading: false,
    path: undefined,
    selectedAction: undefined,
  },
  parameters: {
    name: '',
    typeFilter: QuestTypeFilter.ALL,
    accessFilter: QuestAccessFilter.ALL,
    ironman: false,
    recommended: false,
    lampSkills: [],
    questPriorities: {},
    algorithm: PathFinderAlgorithm.DEFAULT,
  },
  quests: {
    error: false,
    errorResponse: undefined,
    loading: false,
    quests: [],
  },
};

const getters: GetterTree<RootState, RootState> = {
  [constants.GETTERS.GET_FIELD]: getField,
  [constants.GETTERS.GET_QUESTS_PARAMETERS]: (state: RootState) => ({
    name: state.parameters.name,
    typeFilter: state.parameters.typeFilter,
    accessFilter: state.parameters.accessFilter,
  }),
};

const mutations: MutationTree<RootState> = {
  [constants.MUTATIONS.SET_ACTIONS_ERROR](
    state: RootState,
    error: AjaxError
  ): void {
    state.actions.loading = false;
    state.actions.error = true;
    state.actions.errorResponse = error;
  },
  [constants.MUTATIONS.SET_ACTIONS_LOADING](state: RootState): void {
    state.actions.loading = true;
    state.actions.path = undefined;
    state.actions.selectedAction = undefined;
    state.actions.error = false;
  },
  [constants.MUTATIONS.SET_ACTIONS_PATH](state: RootState, path: Path): void {
    path.stats.percentComplete = Math.round(path.stats.percentComplete);

    state.actions.loading = false;
    state.actions.path = path;
    state.actions.selectedAction = head(path.actions);
  },
  [constants.MUTATIONS.SET_PARAMETERS](
    state: RootState,
    parameters: PathFinderParameters
  ): void {
    state.parameters = parameters;
  },
  [constants.MUTATIONS.SET_QUESTS_ERROR](
    state: RootState,
    error: AjaxError
  ): void {
    state.quests.loading = false;
    state.quests.error = true;
    state.quests.errorResponse = error;
  },
  [constants.MUTATIONS.SET_QUESTS_LOADING](state: RootState): void {
    state.quests.loading = true;
    state.actions.error = false;
  },
  [constants.MUTATIONS.SET_QUESTS](state: RootState, quests: Quest[]): void {
    const priorities = state.parameters.questPriorities || {};

    quests = filter(quests, quest => quest.id >= 0);
    quests = map(quests, quest => {
      quest.priority = priorities[quest.id] || QuestPriority.NORMAL;
      return quest;
    });

    state.quests.loading = false;
    state.quests.quests = quests;
  },
  [constants.MUTATIONS.UPDATE_FIELD]: updateField,
};

const actions: ActionTree<RootState, RootState> = {
  async [constants.ACTIONS.FIND_PATH](
    context: ActionContext<RootState, RootState>
  ): Promise<void> {
    context.commit(constants.MUTATIONS.SET_ACTIONS_LOADING);

    const params = context.state.parameters;

    return axios
      .get(PATH_FINDER_URL, {
        params,
        paramsSerializer(params) {
          return qs.stringify(params, {
            arrayFormat: 'repeat',
          });
        },
      })
      .then(response =>
        context.commit(constants.MUTATIONS.SET_ACTIONS_PATH, response.data)
      )
      .catch(error => {
        console.error('Failed to find path', error);
        context.commit(constants.MUTATIONS.SET_ACTIONS_ERROR, {
          error: error,
          parameters: extend({}, params),
        });
      });
  },
  async [constants.ACTIONS.LOAD_QUESTS](
    context: ActionContext<RootState, RootState>
  ): Promise<void> {
    context.commit(constants.MUTATIONS.SET_QUESTS_LOADING);

    const params = context.getters[constants.GETTERS.GET_QUESTS_PARAMETERS];

    return axios
      .get(QUESTS_URL, {
        params,
        paramsSerializer(params) {
          return qs.stringify(params, {
            arrayFormat: 'repeat',
          });
        },
      })
      .then(response => {
        context.commit(constants.MUTATIONS.SET_QUESTS, response.data);
      })
      .catch(error => {
        console.error('Failed to load quests', error);
        context.commit(constants.MUTATIONS.SET_QUESTS_ERROR, {
          error: error,
          parameters: extend({}, params),
        });
        context.state.quests.quests = [];
      });
  },
};

const store = new Store<RootState>({
  plugins,
  state,
  getters,
  mutations,
  actions,
});

export type ComputedMapper<S> = {
  [K in keyof S]: () => S[K];
};

export {RootState, constants, store};
