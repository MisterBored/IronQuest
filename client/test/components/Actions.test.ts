import {createLocalVue, shallowMount} from '@vue/test-utils';
import Vue from 'vue';
import Vuetify from 'vuetify';
import Vuex, {Store} from 'vuex';
import {getField, updateField} from 'vuex-map-fields';

import {QuestAccessFilter, QuestTypeFilter} from '../../src/common/types';
import Actions from '../../src/components/Actions.vue';
import {RootState} from '../../src/store';

Vue.use(Vuetify);

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Actions.vue', () => {
  let store: Store<RootState>;

  beforeEach(() => {
    store = new Store({
      state: {
        actions: {
          error: false,
          errorResponse: undefined,
          loading: false,
          path: undefined,
          selectedAction: undefined,
        },
        parameters: {
          name: '',
          typeFilter: QuestTypeFilter.ALL,
          accessFilter: QuestAccessFilter.ALL,
          ironman: false,
          recommended: false,
          lampSkills: [],
        },
        quests: {
          error: false,
          errorResponse: undefined,
          loading: false,
          quests: [],
        },
      },
      getters: {
        getField,
      },
      mutations: {
        updateField,
      },
    });
  });

  it('should render component and match snapshot', () => {
    const wrapper = shallowMount(Actions, {store, localVue});

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render loading indicator and match snapshot', () => {
    store.state.actions.loading = true;

    const wrapper = shallowMount(Actions, {store, localVue});

    expect(wrapper.html()).toMatchSnapshot();
  });
});
