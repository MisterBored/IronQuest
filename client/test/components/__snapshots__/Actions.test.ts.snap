// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions.vue should render component and match snapshot 1`] = `
<v-container-stub tag="div"
                  fluid="true"
>
  <v-row-stub tag="div"
              align="center"
              justify="center"
  >
    <v-alert-stub tag="div"
                  transition="fade-transition"
                  closelabel="$vuetify.close"
                  dismissible="true"
                  closeicon="$cancel"
                  icon
                  prominent="true"
                  type="error"
    >
      <v-row-stub tag="div"
                  align="center"
      >
        <v-col-stub tag="div"
                    class="grow"
        >
          Something went wrong!
        </v-col-stub>
        <v-col-stub tag="div"
                    class="shrink"
        >
          <v-btn-stub tag="button"
                      activeclass
                      href="https://github.com/darrenswhite/ironquest/issues/new"
                      target="_blank"
                      type="button"
                      rel="noopener noreferrer"
          >
            Submit issue
          </v-btn-stub>
        </v-col-stub>
      </v-row-stub>
    </v-alert-stub>
  </v-row-stub>
</v-container-stub>
`;

exports[`Actions.vue should render loading indicator and match snapshot 1`] = `
<v-container-stub tag="div"
                  fluid="true"
                  fill-height
>
  <v-row-stub tag="div"
              align="center"
              justify="center"
  >
    <v-progress-circular-stub color="primary"
                              indeterminate="true"
                              rotate="0"
                              size="32"
                              width="4"
                              value="0"
    >
    </v-progress-circular-stub>
  </v-row-stub>
</v-container-stub>
`;
